crypto/x509/x509_vfy.c

```diff
@@ -525,6 +526,14 @@ static int check_chain_extensions(X509_STORE_CTX *ctx)
                 ret = 1;
             break;
         }
+        if ((ctx->param->flags & X509_V_FLAG_X509_STRICT) && num > 1) {
+            /* Check for presence of explicit elliptic curve parameters */
+            ret = check_curve(x);
```

Comment:
This assignment seems highly problematic, in that it allows an existing `ret` value of zero (e.g., corresponding to `X509_V_ERR_INVALID_CA` from the previous stanza) to be overwritten with a value of one (success), thus masking the "invalid CA" error.

In master we have avoided this problem by introducing `CB_FAIL_IF()` (originally `CHECK_CB()`) in #13312, that invokes the callback for each potential error and returns immediately, so in the aforementioned case of an invalid CA we don't even get this far.

AFAIK, it is not appropriate to backport #13312 to 1.1.1, so we should probably just skip this check entirely if `ret` is already zero.  I will make a PR.

Reply:
I think this is even CVE worthy.

Reply:
@kaduk, very good that you noticed this!
The problem was not immediate to see when reviewing the insertion of 
```
ret = check_curve(x);
```
because the criticality of masking any `ret == 0` is due to the wider context of this change.

I suggest fixing this simply by replacing
```
        if ((ctx->param->flags & X509_V_FLAG_X509_STRICT) && num > 1) {
```
by
```
        if (err == 0 && num > 1
            && (ctx->param->flags & X509_V_FLAG_X509_STRICT) != 0) {
```

Reply:
For conspiracy theorists & practitioners: 
the bug was committed on 2020-0**9-11** 09:09:29 :wink:

Reply:
I certainly had to be under some alien mind control when committing that buggy change. :rofl:

Reply:
Lessons learnt:
* Thou shalt triple-check (y)our code on 9-11s and Fridays 13th 😜
* Better not release OpenSSL 3.0 on 2021-9-11 -
  otherwise we risk getting subjected to some cyber terrorist act 🤣

Reply:
Getting back to serious mode, what really makes me worry here:
Why is there no negative test case that effectively would have caught this bug?

A first look at `25-test_verify.t` makes one believe that
(even in v1.1.1) there is such a test:
```
ok(!verify("ee-cert", "sslserver", [qw(root-cert)], [qw(ca-nonca)]),
   "fail non-CA untrusted intermediate");
```

Reply:
Ah, now I see that the bug fix added a regression test elsewhere, in `verify_extra_test.c`,
which does catch the issue.

Yet sStill wondering why 
`openssl verify -x509_strict -trusted test/certs/root-cert.pem -untrusted test/certs/ca-nonca.pem test/certs/ee-cert.pem `
does not catch it - there must be a further hidden condition.

Reply:
Meanwhile found why this is so: although 
```
ret = 0
```
had been masked, the subsequent
```
ctx->error = X509_V_ERR_INVALID_CA;
```
was/is not affected, and `apps/verify.c` not only checks the return code but also
`X509_STORE_CTX_get_error(csc)`, which still indicates the error!
Consequently, the `verity` app still says
```
error test/certs/ee-cert.pem: verification failed
```
and returns 2 (a shell error code).

In other words, only library uses of `X509_verify_cert()`
that rely on the returned (zero vs. nonzero) code _alone_ were affected.

Given that `X509_verify_cert()` contains at its end:
```
    /*
     * Safety-net.  If we are returning an error, we must also set ctx->error,
     * so that the chain is not considered verified should the error be ignored
     * (e.g. TLS with SSL_VERIFY_NONE).
     */
    if (ret <= 0 && ctx->error == X509_V_OK)
        ctx->error = X509_V_ERR_UNSPECIFIED;
    return ret;
```
this brings up a idea to make the safety net more tight: 
add
```
    if (ret > 0 && ctx->error != X509_V_OK)
        ret = 0;
```

Reply:
Adding this check, it turns out that in current master 4 test cases start failing,
and for 1.1.1 it's even 51 cases!
This seems to suggest that there are several more such situations than just the one bug discussed and fixed here.
Which is even more worrying.

Reply:
This stuff is a bit hard to debug because the situations occur only in SSL-related tests, 
for which the tests are very involved (if not to say: incomprehensible),
and because the error codes are in part very misleading. 
In particular, `internal_verify()` can throw `X509_V_ERR_UNABLE_TO_VERIFY_LEAF_SIGNATURE` - guess what this means?
It means X509_V_FLAG_PARTIAL_CHAIN is not set, there is only one cert in the chain,  and this cert is not self-issued :-(
To make things worse, `statem_lib.c` translates this to `SSL_AD_UNKNOWN_CA`.

Reply:
It turns out that most of these cases are harmless and simply due to the fact that `verify_cb_cert()` sets `ctx->error` but in case the error is waived by `ctx->verify_cb()` then the original state of `ctx->error` is not restored.